### PR TITLE
feat(wavespeed): add WaveSpeed AI image and video generation

### PR DESCRIPTION
**Pull request by chengzeyi**
## TLDR

Problem this solves:

- WaveSpeed image and video models need hand-rolled REST calls
- No LiteLLM routing, fallbacks, or logging for them
- #34620 covers only the chat half of #34618

How it solves it:

- Adds `wavespeed` image generation with submit-then-poll
- Adds `wavespeed` video generation on the OpenAI video contract
- Registers the OpenAI-compatible chat endpoint as a JSON provider

## User Flow

Before: a developer generating images through the gateway cannot route to WaveSpeed at all, so they call it out of band and lose gateway logging and fallbacks

1. They send POST https://litellm-domain/v1/images/generations with `{"model": "wavespeed/bytedance/seedream-v5.0-pro", "prompt": "a red panda"}`
2. The gateway answers 400 saying the provider is not supported
3. They fall back to calling WaveSpeed themselves, submitting the task and polling for the result in their own code, and nothing about that spend or latency lands in https://litellm-domain/ui/?page=logs

After: the same request works, and the gateway does the polling

1. The proxy admin sets `WAVESPEED_API_KEY` and adds `model: wavespeed/bytedance/seedream-v5.0-pro`
2. The developer sends the same POST https://litellm-domain/v1/images/generations
3. The response is a normal image payload whose `data[0].url` is a hosted WaveSpeed URL, returned once the task finished
4. For video they send POST https://litellm-domain/v1/videos with `{"model": "wavespeed/bytedance/seedance-2.5/text-to-video", "prompt": "..."}`, get back a video id with status `queued`, poll GET https://litellm-domain/v1/videos/{id} until `completed`, then download GET https://litellm-domain/v1/videos/{id}/content
5. Both requests appear in https://litellm-domain/ui/?page=logs

## Relevant issues

Addresses the media generation half of #34618

Relationship to #34620: that PR, opened by a bot and currently blocked on an unsigned CLA, registers the same chat provider and adds 92 LLM cost map entries. This PR deliberately does not duplicate the cost map. It repeats only the small registration needed for `wavespeed` to exist as a provider at all, since the image and video surfaces depend on it. If #34620 lands first I will rebase and drop the overlapping registration hunks; if it does not, this PR still leaves chat working, minus per-model pricing. The chat registration here matches #34620's, including the base URL and the rule that only the leading `wavespeed/` prefix is stripped, and credit for working that out belongs to that PR

Docs live in BerriAI/litellm-docs#962, which has to merge for `code-quality` and `documentation` to go green here: `test_env_keys.py` fails until `WAVESPEED_API_KEY` and `WAVESPEED_API_BASE` are in the environment variables reference

## Pre-Submission checklist

- [x] I have added meaningful tests
- [x] The handful of test files covering my change pass locally, e.g. `uv run pytest tests/test_litellm/<your_test_file>.py -v`
- [x] My PR passes all required CI/CD checks (e.g., lint, schema.d.ts sync check, etc.)
- [x] My PR's scope is as isolated as possible; it only solves 1 specific problem
- [ ] I have received a Greptile **Confidence Score of at least 4/5** before requesting a maintainer review

## Delays in PR merge?

If you're seeing a delay in your PR being merged, ping the LiteLLM Team on [Slack (#pr-review)](https://join.slack.com/t/litellmossslack/shared_invite/zt-3o7nkuyfr-p_kbNJj8taRfXGgQI1~YyA).

## Screenshots / Proof of Fix

Not captured. Every WaveSpeed media call is a billable generation and I did not want to bill a live account from an automated run, so the only runs here are the mocked tests. The commands a reviewer can use once `WAVESPEED_API_KEY` is set, against a proxy configured with `model: wavespeed/wavespeed-ai/z-image/turbo` and `model: wavespeed/bytedance/seedance-2.5/text-to-video`:

```bash
curl http://localhost:4000/v1/images/generations -H "Authorization: Bearer sk-1234" \
  -H "Content-Type: application/json" \
  -d '{"model": "wavespeed/wavespeed-ai/z-image/turbo", "prompt": "a red panda skateboarding", "size": "1024x1024"}'

curl http://localhost:4000/v1/videos -H "Authorization: Bearer sk-1234" \
  -H "Content-Type: application/json" \
  -d '{"model": "wavespeed/bytedance/seedance-2.5/text-to-video", "prompt": "a red panda skateboarding", "seconds": "5"}'
```

## Type

🆕 New Feature

## Caveats (if any)

- Chat cost map entries are left to #34620
- No cost tracking yet for image or video
- Remix, list, delete, edit, extension are not supported
- `response_format: b64_json` is rejected; WaveSpeed returns URLs
- Needs BerriAI/litellm-docs#962 for the docs gates
- Video content downloads go through the shared SSRF guard

### Final Attestation

- [x] The tests check the right things, including the edge cases, and regressions in the respective real-world customer use-cases are not possible after this PR

### CODE REVIEW

**Comment by chengzeyi** [comment]
code-quality and documentation stay red until BerriAI/litellm-docs#962 merges: CI checks out litellm-docs main, which lacks the new env var rows.

File: litellm/llms/wavespeed/videos/transformation.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+    ) -> bytes:
+        output_url: Final = self._extract_output_url(raw_response)
+        httpx_client: Final[HTTPHandler] = _get_httpx_client()
+        video_response: Final = httpx_client.get(output_url)
+        video_response.raise_for_status()
+        return video_response.content
+
+    async def async_transform_video_content_response(
+        self,
+        raw_response: httpx.Response,
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/llms/wavespeed/videos/transformation.py:253]: Fixed in 0286337ce, using `safe_get` and `async_safe_get` from `litellm/litellm_core_utils/url_utils.py` as you suggested, so the download inherits the repo's own validation rather than a bespoke check: the resolved IP is validated and connected to directly, which closes the DNS-rebinding gap, and each redirect hop is re-validated instead of followed blindly. That also let me drop `_get_httpx_client` and `get_async_httpx_client` from this module, so the private-usage suppression is gone too.

I deliberately did not add a WaveSpeed CDN allowlist on top. Output hosts vary by model and region, and a hardcoded list would break `WAVESPEED_API_BASE`-overridden and self-hosted deployments; the shared helper already gives admins `litellm.user_url_allowed_hosts` if they want to narrow it further.

Regression tests are in `TestWaveSpeedVideoContentSSRF`: link-local including `169.254.169.254`, loopback, all three RFC 1918 ranges, IPv6 `::1` and a `file://` URL are rejected on both the sync and async paths, and a public first hop that 302s to the metadata service is rejected on the second hop. Each asserts the internal mock was never called, so a leak fails the test rather than passing quietly. I checked they actually bite by reverting the fix: 11 of the 12 fail against the old code, and the one that passes either way is the "a legitimate public URL still downloads" case.

While in here I also picked up the other review findings on this PR: poll GETs now carry the caller's timeout instead of the client's 600s default, `api_key` and `api_base` reach the handler so a Router supplying either per model works without environment variables, `input_reference` accepts bytes, paths and file handles by inlining them as data URIs since the body is JSON, and `provider_endpoints_support.json` advertises `video_generations`. One of those exposed a real bug your finding led me to: provider resolution hands the media path the chat base URL, which built an unreachable prediction URL, so `get_api_base` now falls back to the media default in that one case while honoring every other override.

Coverage on the wavespeed modules is 100%, and `make lint` passes including the strict, type-discipline and basedpyright budget gates.

File: litellm/llms/wavespeed/image_generation/handler.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        )
+        prediction_id: Final = get_prediction_id(unwrap_envelope(submit_response))
+        result_url: Final = build_result_url(resolved.api_base, prediction_id)
+        deadline: Final = time.time() + DEFAULT_MAX_POLLING_TIME
+        poll_headers: Final = to_request_payload(prepared.headers)
+
+        consecutive_failures = 0  # rebind-ok: counts consecutive poll transport failures
+        while time.time() < deadline:
+            try:
+                poll_response = sync_client.get(url=result_url, headers=poll_headers)
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/llms/wavespeed/image_generation/handler.py]: Fixed in 0286337ce: both poll GETs now pass the caller's timeout instead of the client default. Test asserts the value reaches the request.

File: litellm/llms/wavespeed/videos/transformation.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        seconds: Final = _to_int_seconds(video_create_optional_params.get("seconds"))
+        input_reference: Final = video_create_optional_params.get("input_reference")
+
+        mapped_size: Final = size.lower().replace("x", "*") if isinstance(size, str) and "x" in size.lower() else None
+
+        return to_request_payload(
+            (
+                *((k, v) for k, v in video_create_optional_params.items() if k not in supported),
+                *optional_pair("image", input_reference),
+                *optional_pair("size", mapped_size),
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/llms/wavespeed/videos/transformation.py:143]: Fixed in 0286337ce: bytes, paths, file handles and (filename, content) tuples are inlined as data URIs, since the body is JSON.

File: litellm/images/main.py [hunk 4]
```diff
@@ -405,6 +406,19 @@ def image_generation(
                 timeout=timeout,
                 client=client,
             )
+        elif custom_llm_provider == "wavespeed":
+            return wavespeed_image_generation.image_generation(
+                model=model,
+                prompt=prompt,
+                model_response=model_response,
+                optional_params=optional_params,
+                litellm_params=litellm_params_dict,
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/images/main.py:420]: Fixed in 0286337ce: api_key and api_base are set on litellm_params before dispatch, so Router-supplied per-model values reach the handler.

File: provider_endpoints_support.json [hunk 5]
```diff
@@ -2555,6 +2555,23 @@
...
+        "messages": false,
+        "responses": true,
+        "embeddings": false,
+        "image_generations": true,
+        "audio_transcriptions": false,
+        "audio_speech": false,
+        "moderations": false,
+        "batches": false,
+        "rerank": false,
+        "a2a": false
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply provider_endpoints_support.json]: Fixed in 0286337ce: the entry now sets video_generations true, matching the schema's provider_json_field for the videos endpoint.

File: litellm/llms/wavespeed/videos/transformation.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+        video_id: str,
+        api_base: str,
+        litellm_params: GenericLiteLLMParams,
+        headers: Mapping[str, str],
+        variant: str | None = None,
+    ) -> tuple[str, dict]:  # mutable-ok: base config contract returns bare `dict`
+        return build_result_url(api_base, extract_original_video_id(video_id)), to_request_payload(MappingProxyType({}))
+
+    def _extract_output_url(self, raw_response: httpx.Response) -> str:
+        prediction: Final = unwrap_envelope(raw_response)
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/llms/wavespeed/videos/transformation.py:222]: Not changed: prediction ids are opaque and every video provider behaves this way. Ownership scoping belongs in the shared /videos layer, not here.

File: litellm/images/main.py [hunk 7]
```diff
@@ -405,6 +406,19 @@ def image_generation(
...
+                logging_obj=litellm_logging_obj,
+                timeout=timeout,
+                extra_headers=extra_headers,
+                client=client,
+                aimg_generation=aimg_generation,
+            )
```
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/images/main.py:426]: Matching the black_forest_labs precedent directly above, which also needs a polling handler that the generic image path cannot express. Happy to move it if you prefer.

File: litellm/llms/wavespeed/videos/transformation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chengzeyi) [reply litellm/llms/wavespeed/videos/transformation.py:143]: Follow-up in 301c4c494: recursive_detector rejected the recursion, so the tuple form now unwraps into a flat helper instead.

**Comment by chengzeyi** [comment]
Rebased onto current `litellm_internal_staging` — the branch is `MERGEABLE` again. 64 checks green; the only two red ones are `code-quality` and `documentation`, and they are waiting on BerriAI/litellm-docs#962 rather than on anything in this PR.

**Conflicts** were in three files:

- `constants.py` and `providers.json` — both sides appended to the same lists (cognition / scx-ai upstream, wavespeed here). Kept every entry.
- `get_llm_provider_logic.py` — a real one. Upstream replaced the per-provider hardcoded endpoint checks with the generic `JSONProviderRegistry.get_by_base_url()` lookup, which resolves WaveSpeed from its `providers.json` entry exactly as the hardcoded branch did, and additionally honours an explicitly passed `api_key`. So I **dropped our branch** rather than keeping it — verified with `get_llm_provider(api_base="https://llm.wavespeed.ai/v1")`, which still returns `wavespeed`.

The merge also pulled in `ruff-tests.toml`, whose `PT011` rejects a bare `pytest.raises(ValueError)` in one of our tests. Narrowed it to match the message, which is the better assertion anyway.

**Still blocked on docs.** `code-quality` and `documentation` both fail with:

```
Environment variables read under ./litellm but mentioned nowhere in the docs:
['WAVESPEED_API_BASE', 'WAVESPEED_API_KEY']
```

Both workflows check out `BerriAI/litellm-docs` at its default branch into `docs/my-website` with no `ref`, so they read whatever is on that repo's main. The rows that satisfy this check are in BerriAI/litellm-docs#962, which is open and mergeable. Nothing in this PR can turn those two green — merging #962 first (or in either order) unblocks them.

**Comment by chengzeyi** [comment]
The two red checks here (`code-quality`, `documentation`) are both the same
`documentation_tests/test_env_keys.py` failure, and they can't be fixed from
this PR:

```
Exception: Environment variables read under ./litellm but mentioned nowhere in
the docs: ['WAVESPEED_API_BASE', 'WAVESPEED_API_KEY']
```

Both workflows check out `BerriAI/litellm-docs` at its default branch into
`docs/my-website`, so the test only sees documentation that has already been
merged. The page documenting these two variables is in
**BerriAI/litellm-docs#962** (`docs/providers/wavespeed.md` plus the
`WAVESPEED_API_KEY` / `WAVESPEED_API_BASE` rows in
`docs/proxy/config_settings.md`), which is open, green, and conflict-free.

So the ordering is: merge litellm-docs#962 first, then these checks pass here
on a re-run with no code change. Happy to do it differently if you'd rather
the provider land docs-last — just let me know and I'll adjust.
